diff --git a/driver/android6.0/ioctl_cmd.h b/driver/android6.0/ioctl_cmd.h
index 99836de..650f781 100644
--- a/driver/android6.0/ioctl_cmd.h
+++ b/driver/android6.0/ioctl_cmd.h
@@ -63,6 +63,7 @@
 #define MA_POWER_ON             _IO(MA_IOC_MAGIC, 49)
 #define MA_SID03_REPORT_KEY_ONE        _IO(MA_IOC_MAGIC, 50)
 #define MA_SID03_REPORT_KEY_TWO        _IO(MA_IOC_MAGIC, 51)
+#define HAL_SET_VERSION_INFO           _IOW(MA_IOC_MAGIC,   52, unsigned int)
 
 #endif /* __IOCTL_CMD_H__ */
 
diff --git a/fingerprint/client/libfprint/interface/ca_interface.c b/fingerprint/client/libfprint/interface/ca_interface.c
index 7550eee..a7e79c1 100644
--- a/fingerprint/client/libfprint/interface/ca_interface.c
+++ b/fingerprint/client/libfprint/interface/ca_interface.c
@@ -5,7 +5,6 @@
 #include <errno.h>
 #include <sys/time.h>
 
-
 #include "globalvar.h"
 #include "ca_interface.h"
 #include "process/process.h"
@@ -23,6 +22,7 @@ microarray_properties_t g_config;
 bio_status_t g_bio_status = {0};
 bio_context_t g_bio_context = {0};
 
+void set_version_info_to_driver();
 
 /*hal层的准备工作
  * @return 0:success
@@ -67,6 +67,7 @@ int bio_init(void)
         ma_tac_deinit();
         return -1;
     }
+ 
 
     LOGD("enable fingerprint irq");
     dc_send_cmd(ENABLE_IRQ);
@@ -79,6 +80,9 @@ int bio_init(void)
         LOGE("tac init error, ret=%d", ret);
         return ret;
     }
+
+    set_version_info_to_driver();
+    
     start_process();
     start_process_extern_msg();
     LOGD("setting default active group.");
@@ -91,7 +95,7 @@ int bio_init(void)
     ma_tac_get_energy();//get bkg for factory test get image.
 
     refresh_timer(g_config.is_sensor_changed);
-#if (defined(PRODUCT_E2_OVERSEA) && defined(OEM_KONKA)) || (defined(PRODUCT_DOGE_A163) && defined(OEM_KUSAI))
+#if (defined(PRODUCT_E2_OVERSEA) && defined(OEM_KONKA)) || (defined(PRODUCT_DOGE_A63) && defined(OEM_KUSAI))
     start_screen_status_process();
 #endif
 
@@ -103,8 +107,30 @@ int bio_init(void)
     return SUCCESS;
 }
 
+void set_version_info_to_driver()
+{
+   char version_info[64];
+
+   memset(&version_info, 0, sizeof(version_info));
+   version_info[0] = 1;
+   sprintf(&version_info[1], "afs%d", g_bio_context.chip_id);
+   dc_send_cmd_with_value(HAL_SET_VERSION_INFO, (int*)version_info);
+   memset(&version_info, 0, sizeof(version_info));
+   
+   version_info[0] = 2;
+   sprintf(&version_info[1], "%s_%s_%s_%s",OEM,SUPPORT_SENSOR,BUILD_VERSION,GIT_SHORT_VERSION);
+   dc_send_cmd_with_value(HAL_SET_VERSION_INFO, (int*)version_info);
+   memset(&version_info, 0, sizeof(version_info));
+   
+   version_info[0] = 3;
+   sprintf(&version_info[1], "Microarray");
+   dc_send_cmd_with_value(HAL_SET_VERSION_INFO, (int*)version_info);
+   
+   version_info[0] = 4;
+   sprintf(&version_info[1], "%s-%s", COMPILE_USER, COMPILE_DATE);
+   dc_send_cmd_with_value(HAL_SET_VERSION_INFO, (int*)version_info);
 
-
+}
 
 
 
diff --git a/fingerprint/client/libfprint/process/authenticate.c b/fingerprint/client/libfprint/process/authenticate.c
index 942bc07..8a7ebbc 100644
--- a/fingerprint/client/libfprint/process/authenticate.c
+++ b/fingerprint/client/libfprint/process/authenticate.c
@@ -105,6 +105,9 @@ static int verify(struct cmd_entry *qe)
         }
         if (fid) {
             dc_send_cmd(TIMEOUT_WAKELOCK);
+#if defined(OEM_KUSAI)
+            dc_send_cmd(MA_KEY_F14);
+#endif
             authenticate_notify(cmd_auth->gid, fid, &g_bio_context.auth_token);
             gettimeofday(&now,NULL);
 
diff --git a/fingerprint/client/libfprint/process/enrollcalibrate.c b/fingerprint/client/libfprint/process/enrollcalibrate.c
index c468f1b..f87391f 100644
--- a/fingerprint/client/libfprint/process/enrollcalibrate.c
+++ b/fingerprint/client/libfprint/process/enrollcalibrate.c
@@ -110,7 +110,7 @@ void* enrollcalibrate_work(void * q)
                     	LOGD("do nothing");
                     } else if (ret == ERROR_EXTRACT_FAIL) {
                         LOGD("image extract fail.");
-                        acquired_event_notify(FINGERPRINT_ACQUIRED_INSUFFICIENT);
+                        //acquired_event_notify(FINGERPRINT_ACQUIRED_INSUFFICIENT);
                     } else if (ret == ERROR_ENROLL_DUPLICATE_AREA) {
                     	duplicate_area_num++;
                     	if(duplicate_area_num<6) {
diff --git a/version/product.csv b/version/product.csv
index 58a70d4..4b5a6bb 100644
--- a/version/product.csv
+++ b/version/product.csv
@@ -20,4 +20,4 @@ A156,Yude,A121_N,Trustonic,armeabi arm64-v8a,Android6.0,Yes,Yes,Yes,No,14,3
 Z285,Yude,A121_N,Trustkernel,arm64-v8a,Android6.0,Yes,Yes,Yes,Yes,14,3
 F620,XINGFEI,A83_T,Android,arm64-v8a,Android6.0,Yes,Yes,Yes,No,14,3
 Android,Hct,A120_N,Trustkernel,arm64-v8a armeabi,Android6.0,Yes,Yes,Yes,No,14,3
-DOGE_A63,Kusai,A83_N,Android,arm64-v8a,Android6.0,Yes,Yes,Yes,No,14,3
+DOGE_A63,Kusai,A96_N&A88_N&A64_N,Android,arm64-v8a armeabi,Android6.0,Yes,Yes,Yes,No,14,3
